[quant] Input-Weight Equalization - edge cases

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62367

Summary: Addresses the edge cases where there exists skippable nodes like relu and pooling, and addresses the case where there are nodes that are being quantized but not equalized.

Modifications:
- When we are finding the next equalization scale for additional scaling in the case of connected layers
    - An example is: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv1`), we will traverse through the graph skipping all relu/batchnorm nodes until we reach a call_module node. If it is a quantization observer (`out_quant_obs2`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node. If it is an equalization observer (`inp_eq_obs2`), this means we have two connected layers that need to be equalized.
    - So, we will scale the current op node (`conv1`) with the next equalization scale.

- Finding the previous op node to see if we need to remove the current equalization observer or replace it with a mul node
    - Using the same example as above: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv2`), we will traverse backwards through the graph skipping all relu/batchnorm nodes until we find a call_module node. If it is a quantization observer (`out_quant_obs1`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node.
    - If this node (`conv1`) is being equalized, this means we are in the case where there are two connected layers that are being equalized, so we will remove the current equalization observer.
    - If this node is not being equalized, then we will replace the equalization observer with a mul node like how we did before.

Test Plan: `python test/test_quantization.py TestEqualizeFx`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29989948](https://our.internmc.facebook.com/intern/diff/D29989948)

angelayi committed:
[quant] Input-Weight Equalization - edge cases

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - edge cases"


Summary: Addresses the edge cases where there exists skippable nodes like relu and pooling, and addresses the case where there are nodes that are being quantized but not equalized.

Modifications:
- When we are finding the next equalization scale for additional scaling in the case of connected layers
    - An example is: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv1`), we will traverse through the graph skipping all relu/batchnorm nodes until we reach a call_module node. If it is a quantization observer (`out_quant_obs2`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node. If it is an equalization observer (`inp_eq_obs2`), this means we have two connected layers that need to be equalized.
    - So, we will scale the current op node (`conv1`) with the next equalization scale.

- Finding the previous op node to see if we need to remove the current equalization observer or replace it with a mul node
    - Using the same example as above: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv2`), we will traverse backwards through the graph skipping all relu/batchnorm nodes until we find a call_module node. If it is a quantization observer (`out_quant_obs1`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node.
    - If this node (`conv1`) is being equalized, this means we are in the case where there are two connected layers that are being equalized, so we will remove the current equalization observer.
    - If this node is not being equalized, then we will replace the equalization observer with a mul node like how we did before.

Test Plan: `python test/test_quantization.py TestEqualizeFx`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29989948).

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -318,22 +414,12 @@ def maybe_get_next_input_eq_obs(node: Node, modules: Dict[str, nn.Module]) -> Op
-        if maybe_relu_node is None
-        else maybe_get_next_module(maybe_relu_node, modules, ObserverBase)
-    )
+    maybe_obs_node = maybe_get_following_observer(node, modules, ObserverBase)
```
Why do we need to find `maybe_obs_node` here? Can we directly find `maybe_get_following_observer(node, modules, _InputEqualizationObserver)`?

angelayi replied:
The way the check current works is that it will look for the following call_module node and see if it's the specific given type. However, there should always be a quantization observer in the middle between the op_node and equalization observer, which would result in `maybe_get_following_observer(node, modules, _InputEqualizationObserver)` returning None.

So I first found the quantization observer, and then looked for the equalization observer from the quantization observer.

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -410,7 +496,12 @@ def scale_weight_node(
+    try:
+        next_equalization_scale_reshaped = reshape_scale(next_equalization_scale, 0, scaled_weight)
+    except RuntimeError:
+        raise ValueError(
+            "Input-weight equalization is unable to support general reshapes within models."
+        )
```
Would be useful to add a more descriptive message here saying what the mismatch was

supriyar commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -82,11 +83,20 @@ def _maybe_get_last_node_only_observer(
-        for maybe_obs_node, _ in self.last_node.users.items():
+        last_node = self.last_node
+
+        # Handle the additional mul node that gets added during equalization if
```
Why is this a special case here?

angelayi replied:
The reason why there is a special casing when there is a quantized but not equalized node followed by an equalized node. 

For example, let's say linear1 is not being equalized but linear2 is being equalized. The following graph is what we get after prepare:
```
x -> inp_quant_obs1 -> linear1 -> out_quant_obs1 -> inp_eq_obs2 -> linear2 -> out_quant_obs2
```

After the equalization functions, we have:
```
x -> inp_quant_obs1 -> linear1 -> mul -> out_quant_obs1 (scaled) -> linear2 -> out_quant_obs2
```

In this current function (`maybe_get_last_node_observer`), it originally only looped through the op node's users to find its output quantization observer. However after applying the equalization nodes, linear1's output quantization observer no longer appears directly after linear1, so I added some extra casing.

It should not interfere with the case where the model already has an op node, because in that case I think there would be a output quantization observer inserted before the mul node:
```
x -> inp_quant_obs1 -> linear1 -> out_quant_obs1 -> mul -> out_quant_obs2
```

supriyar replied:
What about the case where linear1 isn't being quantized and we have a mul node after it. Would we end up calling this function in that case?

@jerryzh168 could you help take a look at this change and see if it is okay for the non-equalization use case?

jerryzh168 replied:
right, there is a chance that the mul following linear is not quantized, for example user explicitly configures that to be not quantized through qconfig_dict, I feel this is probably a bit too hacky, let's think about if we have a different workaround

angelayi replied:
> What about the case where linear1 isn't being quantized and we have a mul node after it. Would we end up calling this function in that case?

If linear1 isn't being quantized, then I think it wouldn't enter this function.

> the mul following linear is not quantized, for example user explicitly configures that to be not quantized through qconfig_dict.

In this case, I think the structure of the prepared graph would still be, so the mul would not appear between the linear and output quantization observer if I'm not mistaken:
```
x -> inp_quant_obs1 -> inp_eq_obs -> linear1 -> out_quant_obs1 -> mul
```

jerryzh168 replied:
what if linear is also not quantized and the op following mul is quantized? would we have 
```
linear1 -> mul -> observer -> conv -> ...
```   and when we call `_maybe_get_last_node_only_observer` it would return the observer?

angelayi replied:
I see, I understand that case now. Could we check to see if the mul operator contains a node with the name "equalization_scale" as an argument? Or I could modify the name of the mul node for equalization to be like "mul_for_equalization" and check if the mul operator found contains "equalization" in the name?

supriyar replied:
> Could we check to see if the mul operator contains a node with the name "equalization_scale" as an argument?

This seems okay to me. Can we add an argument to this function to denote equalization is enabled, in which case it checks for this argument for a `mul`?

angelayi committed:
Update on "[quant] Input-Weight Equalization - edge cases"


Summary: Addresses the edge cases where there exists skippable nodes like relu and pooling, and addresses the case where there are nodes that are being quantized but not equalized.

Modifications:
- When we are finding the next equalization scale for additional scaling in the case of connected layers
    - An example is: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv1`), we will traverse through the graph skipping all relu/batchnorm nodes until we reach a call_module node. If it is a quantization observer (`out_quant_obs2`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node. If it is an equalization observer (`inp_eq_obs2`), this means we have two connected layers that need to be equalized.
    - So, we will scale the current op node (`conv1`) with the next equalization scale.

- Finding the previous op node to see if we need to remove the current equalization observer or replace it with a mul node
    - Using the same example as above: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv2`), we will traverse backwards through the graph skipping all relu/batchnorm nodes until we find a call_module node. If it is a quantization observer (`out_quant_obs1`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node.
    - If this node (`conv1`) is being equalized, this means we are in the case where there are two connected layers that are being equalized, so we will remove the current equalization observer.
    - If this node is not being equalized, then we will replace the equalization observer with a mul node like how we did before.

Test Plan: `python test/test_quantization.py TestEqualizeFx`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29989948](https://our.internmc.facebook.com/intern/diff/D29989948)

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - edge cases"


Summary: Addresses the edge cases where there exists skippable nodes like relu and pooling, and addresses the case where there are nodes that are being quantized but not equalized.

Modifications:
- When we are finding the next equalization scale for additional scaling in the case of connected layers
    - An example is: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv1`), we will traverse through the graph skipping all relu/batchnorm nodes until we reach a call_module node. If it is a quantization observer (`out_quant_obs2`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node. If it is an equalization observer (`inp_eq_obs2`), this means we have two connected layers that need to be equalized.
    - So, we will scale the current op node (`conv1`) with the next equalization scale.

- Finding the previous op node to see if we need to remove the current equalization observer or replace it with a mul node
    - Using the same example as above: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv2`), we will traverse backwards through the graph skipping all relu/batchnorm nodes until we find a call_module node. If it is a quantization observer (`out_quant_obs1`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node.
    - If this node (`conv1`) is being equalized, this means we are in the case where there are two connected layers that are being equalized, so we will remove the current equalization observer.
    - If this node is not being equalized, then we will replace the equalization observer with a mul node like how we did before.

Test Plan: `python test/test_quantization.py TestEqualizeFx`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29989948](https://our.internmc.facebook.com/intern/diff/D29989948)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/_equalize.py:
```diff
@@ -298,6 +322,78 @@ def maybe_get_weight_eq_obs_node(op_node: Node, modules: Dict[str, nn.Module]) -
+    temp_node = node
+    while len(temp_node.users.items()) > 0:
+        possible_node = None
+        for user, _ in temp_node.users.items():
```
nit: does `for user in temp_node.users` work?

jerryzh168 commented on torch/quantization/fx/_equalize.py:
```diff
@@ -298,6 +322,78 @@ def maybe_get_weight_eq_obs_node(op_node: Node, modules: Dict[str, nn.Module]) -
+    modules: Dict[str, nn.Module],
+    target_observer_type: Any = object,
+) -> Optional[Node]:
+    """ Loops through the graph forwards and finds the next observer node with
```
if there are multiple paths, looks like this will just follow through one path and find the first observer, might be good to document this I think.

jerryzh168 commented on torch/quantization/fx/_equalize.py:
```diff
@@ -298,6 +322,78 @@ def maybe_get_weight_eq_obs_node(op_node: Node, modules: Dict[str, nn.Module]) -
+
+    temp_node = node
+    while len(temp_node.args) > 0:
+        possible_node = temp_node.args[0]
```
I feel this loop is a little bit confusing, can we just assign `temp_node = temp_node.args[0]`here and redo the condition checks below?

looks like in all three branches we can do `temp_node = temp_node.args[0]`

angelayi committed:
Update on "[quant] Input-Weight Equalization - edge cases"


Summary: Addresses the edge cases where there exists skippable nodes like relu and pooling, and addresses the case where there are nodes that are being quantized but not equalized.

Modifications:
- When we are finding the next equalization scale for additional scaling in the case of connected layers
    - An example is: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv1`), we will traverse through the graph skipping all relu/batchnorm nodes until we reach a call_module node. If it is a quantization observer (`out_quant_obs2`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node. If it is an equalization observer (`inp_eq_obs2`), this means we have two connected layers that need to be equalized.
    - So, we will scale the current op node (`conv1`) with the next equalization scale.

- Finding the previous op node to see if we need to remove the current equalization observer or replace it with a mul node
    - Using the same example as above: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv2`), we will traverse backwards through the graph skipping all relu/batchnorm nodes until we find a call_module node. If it is a quantization observer (`out_quant_obs1`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node.
    - If this node (`conv1`) is being equalized, this means we are in the case where there are two connected layers that are being equalized, so we will remove the current equalization observer.
    - If this node is not being equalized, then we will replace the equalization observer with a mul node like how we did before.

Test Plan: `python test/test_quantization.py TestEqualizeFx`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29989948](https://our.internmc.facebook.com/intern/diff/D29989948)

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - edge cases"


Summary: Addresses the edge cases where there exists skippable nodes like relu and pooling, and addresses the case where there are nodes that are being quantized but not equalized.

Modifications:
- When we are finding the next equalization scale for additional scaling in the case of connected layers
    - An example is: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv1`), we will traverse through the graph skipping all relu/batchnorm nodes until we reach a call_module node. If it is a quantization observer (`out_quant_obs2`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node. If it is an equalization observer (`inp_eq_obs2`), this means we have two connected layers that need to be equalized.
    - So, we will scale the current op node (`conv1`) with the next equalization scale.

- Finding the previous op node to see if we need to remove the current equalization observer or replace it with a mul node
    - Using the same example as above: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv2`), we will traverse backwards through the graph skipping all relu/batchnorm nodes until we find a call_module node. If it is a quantization observer (`out_quant_obs1`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node.
    - If this node (`conv1`) is being equalized, this means we are in the case where there are two connected layers that are being equalized, so we will remove the current equalization observer.
    - If this node is not being equalized, then we will replace the equalization observer with a mul node like how we did before.

Test Plan: `python test/test_quantization.py TestEqualizeFx`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29989948](https://our.internmc.facebook.com/intern/diff/D29989948)

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - edge cases"


Summary: Addresses the edge cases where there exists skippable nodes like relu and pooling, and addresses the case where there are nodes that are being quantized but not equalized.

Modifications:
- When we are finding the next equalization scale for additional scaling in the case of connected layers
    - An example is: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv1`), we will traverse through the graph skipping all relu/batchnorm nodes until we reach a call_module node. If it is a quantization observer (`out_quant_obs2`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node. If it is an equalization observer (`inp_eq_obs2`), this means we have two connected layers that need to be equalized.
    - So, we will scale the current op node (`conv1`) with the next equalization scale.

- Finding the previous op node to see if we need to remove the current equalization observer or replace it with a mul node
    - Using the same example as above: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv2`), we will traverse backwards through the graph skipping all relu/batchnorm nodes until we find a call_module node. If it is a quantization observer (`out_quant_obs1`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node.
    - If this node (`conv1`) is being equalized, this means we are in the case where there are two connected layers that are being equalized, so we will remove the current equalization observer.
    - If this node is not being equalized, then we will replace the equalization observer with a mul node like how we did before.

Test Plan: `python test/test_quantization.py TestEqualizeFx`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29989948](https://our.internmc.facebook.com/intern/diff/D29989948)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29989948).

angelayi committed:
Update on "[quant] Input-Weight Equalization - edge cases"


Summary: Addresses the edge cases where there exists skippable nodes like relu and pooling, and addresses the case where there are nodes that are being quantized but not equalized.

Modifications:
- When we are finding the next equalization scale for additional scaling in the case of connected layers
    - An example is: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv1`), we will traverse through the graph skipping all relu/batchnorm nodes until we reach a call_module node. If it is a quantization observer (`out_quant_obs2`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node. If it is an equalization observer (`inp_eq_obs2`), this means we have two connected layers that need to be equalized.
    - So, we will scale the current op node (`conv1`) with the next equalization scale.

- Finding the previous op node to see if we need to remove the current equalization observer or replace it with a mul node
    - Using the same example as above: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv2`), we will traverse backwards through the graph skipping all relu/batchnorm nodes until we find a call_module node. If it is a quantization observer (`out_quant_obs1`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node.
    - If this node (`conv1`) is being equalized, this means we are in the case where there are two connected layers that are being equalized, so we will remove the current equalization observer.
    - If this node is not being equalized, then we will replace the equalization observer with a mul node like how we did before.

Test Plan: `python test/test_quantization.py TestEqualizeFx`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29989948](https://our.internmc.facebook.com/intern/diff/D29989948)

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - edge cases"


Summary: Addresses the edge cases where there exists skippable nodes like relu and pooling, and addresses the case where there are nodes that are being quantized but not equalized.

Modifications:
- When we are finding the next equalization scale for additional scaling in the case of connected layers
    - An example is: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv1`), we will traverse through the graph skipping all relu/batchnorm nodes until we reach a call_module node. If it is a quantization observer (`out_quant_obs2`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node. If it is an equalization observer (`inp_eq_obs2`), this means we have two connected layers that need to be equalized.
    - So, we will scale the current op node (`conv1`) with the next equalization scale.

- Finding the previous op node to see if we need to remove the current equalization observer or replace it with a mul node
    - Using the same example as above: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv2`), we will traverse backwards through the graph skipping all relu/batchnorm nodes until we find a call_module node. If it is a quantization observer (`out_quant_obs1`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node.
    - If this node (`conv1`) is being equalized, this means we are in the case where there are two connected layers that are being equalized, so we will remove the current equalization observer.
    - If this node is not being equalized, then we will replace the equalization observer with a mul node like how we did before.

Test Plan: `python test/test_quantization.py TestEqualizeFx`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29989948](https://our.internmc.facebook.com/intern/diff/D29989948)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29989948).

supriyar commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -1607,7 +1638,8 @@ def convert(self,
                 is_reference: bool = False,
-                convert_custom_config_dict: Dict[str, Any] = None) -> Node:
+                convert_custom_config_dict: Dict[str, Any] = None,
+                equalization_enabled: bool = False) -> Node:
```
Do we need to add this argument to add the instances of `convert`, given that it has a default value?

supriyar reviewed (approved):
Looks good, please take a look at the comment before landing

angelayi committed:
Update on "[quant] Input-Weight Equalization - edge cases"


Summary: Addresses the edge cases where there exists skippable nodes like relu and pooling, and addresses the case where there are nodes that are being quantized but not equalized.

Modifications:
- When we are finding the next equalization scale for additional scaling in the case of connected layers
    - An example is: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv1`), we will traverse through the graph skipping all relu/batchnorm nodes until we reach a call_module node. If it is a quantization observer (`out_quant_obs2`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node. If it is an equalization observer (`inp_eq_obs2`), this means we have two connected layers that need to be equalized.
    - So, we will scale the current op node (`conv1`) with the next equalization scale.

- Finding the previous op node to see if we need to remove the current equalization observer or replace it with a mul node
    - Using the same example as above: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv2`), we will traverse backwards through the graph skipping all relu/batchnorm nodes until we find a call_module node. If it is a quantization observer (`out_quant_obs1`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node.
    - If this node (`conv1`) is being equalized, this means we are in the case where there are two connected layers that are being equalized, so we will remove the current equalization observer.
    - If this node is not being equalized, then we will replace the equalization observer with a mul node like how we did before.

Test Plan: `python test/test_quantization.py TestEqualizeFx`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29989948](https://our.internmc.facebook.com/intern/diff/D29989948)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29989948).

angelayi committed:
Update on "[quant] Input-Weight Equalization - edge cases"


Summary: Addresses the edge cases where there exists skippable nodes like relu and pooling, and addresses the case where there are nodes that are being quantized but not equalized.

Modifications:
- When we are finding the next equalization scale for additional scaling in the case of connected layers
    - An example is: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv1`), we will traverse through the graph skipping all relu/batchnorm nodes until we reach a call_module node. If it is a quantization observer (`out_quant_obs2`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node. If it is an equalization observer (`inp_eq_obs2`), this means we have two connected layers that need to be equalized.
    - So, we will scale the current op node (`conv1`) with the next equalization scale.

- Finding the previous op node to see if we need to remove the current equalization observer or replace it with a mul node
    - Using the same example as above: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv2`), we will traverse backwards through the graph skipping all relu/batchnorm nodes until we find a call_module node. If it is a quantization observer (`out_quant_obs1`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node.
    - If this node (`conv1`) is being equalized, this means we are in the case where there are two connected layers that are being equalized, so we will remove the current equalization observer.
    - If this node is not being equalized, then we will replace the equalization observer with a mul node like how we did before.

Test Plan: `python test/test_quantization.py TestEqualizeFx`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29989948](https://our.internmc.facebook.com/intern/diff/D29989948)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29989948).

jerryzh168 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -76,17 +77,33 @@ def __init__(self, node: Node, modules: Dict[str, torch.nn.Module]):
+                # Check that the mul node contains an 'equalization_scale' as an
+                # argument to ensure that this mul node is used for equalization.
+                for arg in maybe_eq_node.args:
+                    if isinstance(arg, Node) and "equalization_scale" in arg.name:
```
actually Node name is a fragile thing to check I think, is it possible to check for something else?

angelayi replied:
Could we check with node.target?

angelayi committed:
Update on "[quant] Input-Weight Equalization - edge cases"


Summary: Addresses the edge cases where there exists skippable nodes like relu and pooling, and addresses the case where there are nodes that are being quantized but not equalized.

Modifications:
- When we are finding the next equalization scale for additional scaling in the case of connected layers
    - An example is: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv1`), we will traverse through the graph skipping all relu/batchnorm nodes until we reach a call_module node. If it is a quantization observer (`out_quant_obs2`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node. If it is an equalization observer (`inp_eq_obs2`), this means we have two connected layers that need to be equalized.
    - So, we will scale the current op node (`conv1`) with the next equalization scale.

- Finding the previous op node to see if we need to remove the current equalization observer or replace it with a mul node
    - Using the same example as above: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv2`), we will traverse backwards through the graph skipping all relu/batchnorm nodes until we find a call_module node. If it is a quantization observer (`out_quant_obs1`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node.
    - If this node (`conv1`) is being equalized, this means we are in the case where there are two connected layers that are being equalized, so we will remove the current equalization observer.
    - If this node is not being equalized, then we will replace the equalization observer with a mul node like how we did before.

Test Plan: `python test/test_quantization.py TestEqualizeFx`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29989948](https://our.internmc.facebook.com/intern/diff/D29989948)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29989948).

supriyar commented:
@angelayi is the mypy failure related to the changes?

angelayi committed:
Update on "[quant] Input-Weight Equalization - edge cases"


Summary: Addresses the edge cases where there exists skippable nodes like relu and pooling, and addresses the case where there are nodes that are being quantized but not equalized.

Modifications:
- When we are finding the next equalization scale for additional scaling in the case of connected layers
    - An example is: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv1`), we will traverse through the graph skipping all relu/batchnorm nodes until we reach a call_module node. If it is a quantization observer (`out_quant_obs2`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node. If it is an equalization observer (`inp_eq_obs2`), this means we have two connected layers that need to be equalized.
    - So, we will scale the current op node (`conv1`) with the next equalization scale.

- Finding the previous op node to see if we need to remove the current equalization observer or replace it with a mul node
    - Using the same example as above: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv2`), we will traverse backwards through the graph skipping all relu/batchnorm nodes until we find a call_module node. If it is a quantization observer (`out_quant_obs1`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node.
    - If this node (`conv1`) is being equalized, this means we are in the case where there are two connected layers that are being equalized, so we will remove the current equalization observer.
    - If this node is not being equalized, then we will replace the equalization observer with a mul node like how we did before.

Test Plan: `python test/test_quantization.py TestEqualizeFx`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29989948](https://our.internmc.facebook.com/intern/diff/D29989948)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29989948).
